### DISCUSSION TITLE
Adding Local Outlier Factor (LOF) algorithms

### DISCUSSION THREAD

**Discussion by alexbeattie42** [online-ml/river#706] [Ideas]
## Overview
I am working on my Master's Thesis related to outlier detection in streaming data. The only anomaly detection method I've found so far in river is [half space trees](https://riverml.xyz/latest/api/anomaly/HalfSpaceTrees/). Per the documentation “Half space trees work well when anomalies are spread out but do not work well if anomalies are packed together in windows.”

## Problem
 I have been reading on various Local Outlier Factor (LOF) implementations for streaming data. [1](https://dl.acm.org/doi/pdf/10.1145/3219819.3220022) and [2](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7085525/) seem fairly promising. Adding an LOF algorithm to improve their anomaly detection is on the river roadmap already so it seems like a [desired feature](https://www.notion.so/d1e86fcdf21e4deda16eedab2b3361fb?v=503f44740b8b44a99a961aa96e9e46e1&p=f718f43301624532af9cc2498f66cf62).

## Proposal
Use a few standard streaming benchmark datasets as well as my group's private data to perform some tests.

### Methods
For the first set of tests I plan to include existing streaming methods (ex. river’s half space trees outlier detection method, [pysad](https://pysad.readthedocs.io/en/latest/api.htmland), etc.) Then I plan to compare those results to results obtained by running the entire dataset through a non-streaming approach (ex. [pyOD methods](https://pyod.readthedocs.io/en/latest/pyod.html) , statistical methods, etc) to see if the current streaming approaches provides comparable results (my hypothesis is that they will not for certain cases, leading to the next part).

### Addition to River
Implement one of the proposed LOF variants referenced in the "problem" section in river and compare it to the current half space tree method or other methods from pySAD. 

## Question

Before starting this work, I wanted to make sure that adding the LOF streaming method is in line with the goals of the library and would be something useful. Any ideas or suggestions on good benchmark datasets (I'm currently compiling a list of commonly used ones in the literature) or methodology improvements are also welcome.

**Comment by MaxHalford** [comment]
Hello, thank you for your motivation

> Before starting this work, I wanted to make sure that adding the LOF streaming method is in line with the goals of the library and would be something useful.

I see no problem. In fact, online anomaly detection is in demand, so a contribution would be very welcome.

The first step is too see to what extend LOF can be performed online. We don't want to have semi-batch approaches that aren't truly online. Ideally, the complexity of `learn_one` should be O(1).

> Any ideas or suggestions on good benchmark datasets (I'm currently compiling a list of commonly used ones in the literature) or methodology improvements are also welcome.

I'm sure you'll find the "good" ones by yourself. There aren't that many. If you do find one that we don't have, feel free to add it to the `datasets` module.

*Meta: you can make use of Markdown syntax to insert links. You don't have to copy/paste in plain text.*

**Reply by ghost** [reply]
Alex - [Kate Highnam's BETH dataset](https://www.kaggle.com/katehighnam/beth-dataset) is a recently created dataset (2021) containing anomalies from host and network cyber logs. The dataset is labelled with suspicious and malicious fields - very useful.

**Reply by MaxHalford** [reply]
That looks good! Does anyone feel confident enough to add it to the datasets module? I can review the code of course.

**Reply by alexbeattie42** [reply]
I can try to add the dataset. Since Kaggle requires a user to be logged in before downloading the dataset will have to be hosted elsewhere. I'm seeing many of the datasets are hosted (via GitHub pages I believe) at "https://maxhalford.github.io/files/datasets/*". Should this one also be hosted there?

**Reply by MaxHalford** [reply]
Good question! Maybe we should look into hosting the datasets on a university host. @jacobmontiel maybe we could use Polytechnique or Waikato to host datasets?

**Reply by jinxmirror13** [reply]
Happy to add the dataset to Google datasets (which I hope you don't need a login for!) or another site if you can recommend one... Apologies, I did not realise Kaggle required a user account!

**Reply by alexbeattie42** [reply]
If I understand correctly, google datasets is primarily a search aggregator for different hosting sources and doesn't allow you to directly host a dataset with them. After some preliminary searching the [UCI Machine Learning Archive](https://archive-beta.ics.uci.edu/ml/donation-policy) was the only reputable public host I could find for "donating/open-sourcing" a dataset that does not implement a login/registration wall for downloading a dataset. 

It could also be hosted privately via a variety of methods but they would then need to be maintained by a private individual or university which could present difficulties in the long term

**Reply by MaxHalford** [reply]
The UCI machine learning learning archive sounds like a good idea. I don't why I didn't think about it before! Thanks for looking into it.

Hey @jinxmirror13, how do you feel about adding it there?

**Reply by jinxmirror13** [reply]
No problems publishing it elsewhere so more users can access it! I ran into some issues trying to publish it with them, but reaching out to see what can be done. Thank you for the recommendation @alexbeattie42 and @MaxHalford !

**Reply by jinxmirror13** [reply]
So I didn't manage to get UCI to work, but discovered an internal service at Imperial which did. You should be able to get the current version of the full dataset by running this:

```
wget https://data.hpc.imperial.ac.uk/resolve/\?doi\=9422\&file\=4\&access\= -O full_BETH_dataset.zip
```

**Reply by MaxHalford** [reply]
That's great, thanks a lot of. I'll add the dataset very soon. For my interest, is that your university of affiliation, or can anyone theoretically store datasets there?

**Reply by jinxmirror13** [reply]
You have to be affiliated, but maybe other Universities are starting to offer similar services?

**Reply by MaxHalford** [reply]
Yes, you're right. It's just something I haven't looked into yet. Thanks again! I'll ping everyone in this discussion once I've added the dataset.

Edit: the dataset is quite complex, and there are different ways to process it. I'll have to take some more time later to add it!
